mock-crust feature tests

brian-js opened the pull request:
Include stress tests using the mock-crust feature.

Added one test that randomly stores structured or immutable data to the network checking the relevant number of copies are stored when repeatedly dropping three nodes then adding three nodes. 
Another test done by @maqi has been merged to this PR that checks data availability in the presence of churn during storage and retrieval of immutable data.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/maidsafe/safe_vault/401) &emsp; Multiple assignees:&emsp;<img alt="@Fraser999" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/190532?s=40&v=3">&nbsp;<a href="/maidsafe/safe_vault/pulls/assigned/Fraser999">Fraser999</a>,&emsp;<img alt="@dirvine" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/123627?s=40&v=3">&nbsp;<a href="/maidsafe/safe_vault/pulls/assigned/dirvine">dirvine</a>,&emsp;<img alt="@madadam" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/5271?s=40&v=3">&nbsp;<a href="/maidsafe/safe_vault/pulls/assigned/madadam">madadam</a>
<!-- Reviewable:end -->

maidsafe-highfive commented:
r? @Fraser999

(maidsafe_highfive has picked a reviewer for you, use r? to override)

brian-js commented:
+@dirvine +@madadam +@fraser999 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-:-KEk0hpnybIjwzNzcQY3:-570793504)*
<!-- Sent from Reviewable.io -->

madadam commented on src/tests/mod.rs:
```diff
@@ -52,3 +171,193 @@ fn plain_data_put_and_get() {
+    let put_requests = put_range.ind_sample(&mut rng);
+
+    for _ in 0..put_requests {
+        if random::<usize>() % 2 == 0 {
```
`if random::<bool>() {` is cleaner I think.

madadam commented on src/tests/mod.rs:
```diff
@@ -20,12 +20,131 @@
+}
 
 #[test]
 fn plain_data_put_and_get() {
```
This function should be called `immutable_data_put_and_get` (it was my mistake, but I noticed it just now).

dirvine commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/tests/mod.rs, line 24 \[r6\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEkVWE9tn_0eeRCv8A8:-KEkVWE9tn_0eeRCv8A9:-899749387) ([raw file](https://github.com/maidsafe/safe_vault/blob/6bf9991bd2768ac9d04830992b6c5e70bc5413ab/src/tests/mod.rs#L24)):*
Will let this go for now, but normally we use the last part of the namespace when accessing functions. So `routing::normal_to_backup` is the way to call the function. For types then this is OK, just not methods. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

dirvine commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/tests/mod.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEkWd3S4O_HOPTDYwOC:-KEkWd3S4O_HOPTDYwOD:-1390248334) ([raw file](https://github.com/maidsafe/safe_vault/blob/ec3b13f74c55f1d4c71796a6f5340115d428966e/src/tests/mod.rs#L289)):*
Do we need this with mock crust? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

madadam commented:
*[src/tests/mod.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEkWd3S4O_HOPTDYwOC:-KEkX-r5W3rd3twE0SNm:1824965665) ([raw file](https://github.com/maidsafe/safe_vault/blob/ec3b13f74c55f1d4c71796a6f5340115d428966e/src/tests/mod.rs#L289)):*
We definitely shouldn't need it, unless there is a bug somewhere. I'm currently investigating it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

dirvine commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/tests/mod.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEkWd3S4O_HOPTDYwOC:-KEkY_pAJAAfH1bW6kyt:-542661589) ([raw file](https://github.com/maidsafe/safe_vault/blob/ec3b13f74c55f1d4c71796a6f5340115d428966e/src/tests/mod.rs#L289)):*
Excellent @madadam  It does seem superfluous or should be, if it's removed please remove the include as well (time::Duration)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

brian-js commented:
*[src/tests/mod.rs, line 289 \[r6\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEkWd3S4O_HOPTDYwOC:-KEkpBDzBSRi1t2F8PBr:1835837281) ([raw file](https://github.com/maidsafe/safe_vault/blob/ec3b13f74c55f1d4c71796a6f5340115d428966e/src/tests/mod.rs#L289)):*
Removed and updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

brian-js commented:
*[src/tests/mod.rs, line 197 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEaEHGbN4u1XLRpd0JC-r5-197:-KEkpI4QlWRELgs1-kqk:-499146009) ([raw file](https://github.com/maidsafe/safe_vault/blob/c7203e832b64a069e9475f606e1a1a524eb73a51/src/tests/mod.rs#L197)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

brian-js commented:
*[src/tests/mod.rs, line 150 \[r5\]](https://reviewable.io:443/reviews/maidsafe/safe_vault/401#-KEaEHGbN4u1XLRpd0JC-r5-150:-KEkpKTZkk4gY4J_fwwg:-499146009) ([raw file](https://github.com/maidsafe/safe_vault/blob/c7203e832b64a069e9475f606e1a1a524eb73a51/src/tests/mod.rs#L150)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->

dirvine commented:
Reviewed 2 of 4 files at r5, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_vault/401)*
<!-- Sent from Reviewable.io -->
